agent/engine/docker_task_engine.go

```diff
@@ -274,48 +278,31 @@ func (engine *DockerTaskEngine) monitorExecAgentProcesses(ctx context.Context) {
 	defer engine.tasksLock.RUnlock()
 	for _, mTask := range engine.managedTasks {
 		task := mTask.Task
-		if !task.IsExecCommandAgentEnabled() {
+		if !task.IsExecCommandAgentEnabled() || task.GetKnownStatus() != apitaskstatus.TaskRunning {
 			continue
 		}
 		for _, c := range task.Containers {
-			go engine.monitorExecAgentRunning(ctx, task.Arn, c)
+			go engine.monitorExecAgentRunning(ctx, task, c)
 		}
 	}
 }
 
 func (engine *DockerTaskEngine) monitorExecAgentRunning(ctx context.Context,
-	tArn string, c *apicontainer.Container) {
-	if c.GetExecCommandAgentMetadata() == nil {
+	task *apitask.Task, c *apicontainer.Container) {
+	if !c.IsRunning() {
 		return
 	}
-	pID := c.GetExecCommandAgentMetadata().PID
-	cName := c.Name
-	cID := c.RuntimeID
-	if cID == "" || pID == "" {
-		return
-	}
-	resp, err := engine.client.TopContainer(ctx, cID, dockerclient.TopContainerTimeout, pID)
+	dockerID, err := engine.getDockerID(task, c)
 	if err != nil {
-		if _, ok := err.(*dockerapi.DockerTimeoutError); ok {
-			seelog.Warnf("Task engine [%s]: Timed out monitoring ExecCommandAgent process for container %s", tArn, cName)
-			return
-		}
-		if strings.Contains(err.Error(), dockerapi.TopProcessNotFoundErrorName) {
-			seelog.Errorf("Task engine [%s]: ExecCommandAgent Process for container %s not running, restarting", tArn, cName)
-			c.SetExecCommandAgentMetadata(&apicontainer.ExecCommandAgentMetadata{})
-			// TODO: [ecs-exec]restart exec agent process for this container
-			return
-		}
-		// TODO: [ecs-exec] consider retry/adding an extra docker exec inspect API call to get more info
-		seelog.Errorf("Task engine [%s]: Error finding ExecCommandAgent process for container %s: %v", tArn, cName, err)
+		seelog.Errorf("Task engine [%s]: Could not retrieve docker id for container", task.Arn)
 		return
 	}
-	// no error, but SSM agent process not found in the container
-	if resp == nil || len(resp.Processes) == 0 {
-		seelog.Errorf("Task engine [%s]: ExecCommandAgent Process not found for container %s, no error, restarting", tArn, cName)
-		c.SetExecCommandAgentMetadata(&apicontainer.ExecCommandAgentMetadata{})
-		// TODO: [ecs-exec]restart exec agent process for this container
+	_, err = engine.execCmdAgentMgr.RestartIfStopped(ctx, engine.client, task, c, dockerID)
+	if err != nil {
+		seelog.Errorf("Task engine [%s]: Failed to restart ExecCommandAgent Process for container [%s]: %v", task.Arn, dockerID, err)
```

Comment:
on restart failing, the exec metadata needs to be reset https://github.com/aws/amazon-ecs-agent/pull/2648/files#diff-b9aa12c3425c34d6f330f28407f779f6L305

Reply:
Good catch, I will reset it inside `RestartIfStopped` because we still need the metadata in the following scenario:
* Docker exec inspect fails: in this case we could not verify if the agent is running, in the next health check we need the previous metadata to do another docker exec inspect, if we remove it, we would lose track of it and it might still be running.

Reply:
Addresssed